lib/auth/userloginstate/generator_test.go

```diff
@@ -0,0 +1,254 @@
+					"trait2": []string{"value3"},
+				}),
+			},
+			roles: []string{"role1", "role2"},
+			expected: newUserLoginState(t, "user",
+				[]string{
+					"role1",
+				}, trait.Traits{
+					"trait1": []string{"value1"},
+				}),
+		},
+		{
+			name: "access lists add roles with duplicates",
+			accessLists: []*accesslist.AccessList{
+				newAccessList(t, "1", []string{"user"}, []string{"role1", "role2"}, trait.Traits{}),
+				newAccessList(t, "2", []string{"user"}, []string{"role2", "role3"}, trait.Traits{}),
+			},
+			roles: []string{"role1", "role2", "role3"},
+			expected: newUserLoginState(t, "user",
+				[]string{
+					"role1",
+					"role2",
+					"role3",
+				}, trait.Traits{}),
+		},
+		{
+			name: "access lists add traits with duplicates",
+			accessLists: []*accesslist.AccessList{
+				newAccessList(t, "1", []string{"user"}, []string{},
+					trait.Traits{
+						"trait1": []string{"value1", "value2"},
+						"trait2": []string{"value3", "value4"},
+					},
+				),
+				newAccessList(t, "2", []string{"user"}, []string{},
+					trait.Traits{
+						"trait2": []string{"value3", "value1"},
+						"trait3": []string{"value5", "value6"},
+					},
+				),
+			},
+			expected: newUserLoginState(t, "user", []string{},
+				trait.Traits{
+					"trait1": []string{"value1", "value2"},
+					"trait2": []string{"value3", "value4", "value1"},
+					"trait3": []string{"value5", "value6"},
+				}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+			svc, backendSvc := initSvc(t)
+
+			for _, accessList := range test.accessLists {
+				_, err = backendSvc.UpsertAccessList(ctx, accessList)
+				require.NoError(t, err)
+			}
+
+			for _, role := range test.roles {
+				role, err := types.NewRole(role, types.RoleSpecV6{})
+				require.NoError(t, err)
+				require.NoError(t, backendSvc.UpsertRole(ctx, role))
+			}
+
+			state, err := svc.Generate(ctx, user)
+			require.NoError(t, err)
+			require.Empty(t, cmp.Diff(test.expected, state,
+				cmpopts.SortSlices(func(str1, str2 string) bool {
+					return str1 < str2
+				})))
+		})
+	}
+}
+
+type svc struct {
+	services.AccessLists
+	services.Access
+}
+
+func initSvc(t *testing.T) (*Generator, *svc) {
+	t.Helper()
+
+	clock := clockwork.NewFakeClock()
+	mem, err := memory.New(memory.Config{
+		Clock: clock,
+	})
+	require.NoError(t, err)
+
+	accessListsSvc, err := local.NewAccessListService(mem, clock)
+	require.NoError(t, err)
+	accessSvc := local.NewAccessService(mem)
+
+	return &Generator{accessLists: accessListsSvc, roles: accessSvc, clock: clock},
+		&svc{AccessLists: accessListsSvc, Access: accessSvc}
+}
+
+func newAccessList(t *testing.T, name string, members []string, roles []string, traits trait.Traits) *accesslist.AccessList {
+	t.Helper()
+
+	alMembers := make([]accesslist.Member, len(members))
+	for i, member := range members {
+		alMembers[i] = accesslist.Member{
+			Name:    member,
+			Joined:  time.Now(),
```

Comment:
This is strictly outside of scope of this particular PR, but I wonder if we could make the access lists more flexible. Right now we have:

```
// IsMember will return true if the user is a member for the current list.
func IsMember(identity tlsca.Identity, clock clockwork.Clock, accessList *accesslist.AccessList) error {
	username := identity.Username
	for _, member := range accessList.Spec.Members {
		if member.Name == username && clock.Now().Before(member.Expires) {
			if !userMeetsRequirements(identity, accessList.Spec.MembershipRequires) {
				return trace.AccessDenied("user %s does not meet membership requirements", username)
			}
			return nil
		}
	}

	return trace.NotFound("user %s is not a member of the access list", username)
}
```

If a person is a member of the access list, they are on it until their membership expires. And this is it. 

We could make the membership more flexible, though: for example, based on time of day or day of week, or any other flexible schedule. In the simplest case you should be able to issue membership in the future and have it only valid after given point of time.

Reply:
Interesting. That's definitely something we can look into in the future! It would make for some interesting use cases.

Reply:
Yep, shift-based entitlements come to mind in particular.